### PR TITLE
Doc updates for Python streaming

### PR DESCRIPTION
**Pull request by melap**
In-progress draft for initial review. Still needs snippet URLs

### CODE REVIEW

**Comment by melap** [comment]
Staged: http://apache-beam-website-pull-requests.storage.googleapis.com/410/index.html

File: src/documentation/sdks/python-streaming.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+layout: section
+title: "Apache Beam Python Streaming Pipelines"
+section_menu: section-menu/sdks.html
+permalink: /documentation/sdks/python-streaming/
+---
+
+# Python Streaming Pipelines
+
+Apache Beam SDK for Python supports streaming pipeline execution as of version
+TBD. Currently, two Beam runners support Python streaming execution:
```
Comment (aaltay) [inline src/documentation/sdks/python-streaming.md]: 2 beam runners are the only two beam python runners anyway. I do not know if this is worth mentioning.

File: src/documentation/sdks/python-streaming.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
+---
+layout: section
+title: "Apache Beam Python Streaming Pipelines"
+section_menu: section-menu/sdks.html
+permalink: /documentation/sdks/python-streaming/
+---
+
+# Python Streaming Pipelines
+
+Apache Beam SDK for Python supports streaming pipeline execution as of version
```
Comment (aaltay) [inline src/documentation/sdks/python-streaming.md]: DirectRunner can do this since 2.1.0 and DataflowRunner will start in 2.5.0. Can we try to rephrase as this is experimentally available with some limitations starting from 2.5.0 ?

File: src/documentation/sdks/python-streaming.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+TBD. Currently, two Beam runners support Python streaming execution:
+[DirectRunner]({{ site.baseurl }}/documentation/runners/direct/#streaming-execution)
+and [DataflowRunner]({{ site.baseurl }}/documentation/runners/dataflow/#streaming-execution).
+
+
+## Why use streaming execution?
+
+Beam creates an unbounded PCollection if your pipeline reads from a streaming or
+continously-updating data source (such as Cloud Pub/Sub or Kafka). A runner must
```
Comment (aaltay) [inline src/documentation/sdks/python-streaming.md]: Kafka is not supported, I will drop it from examples.

File: src/documentation/sdks/python-streaming.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+```
+
+These streaming WordCount snippets are from
+[streaming_wordcount.py](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/examples/streaming_wordcount.py).
+This code uses an I/O connector that reads from and writes to an unbounded
+source (Cloud Pub/Sub) and specifies a fixed windowing strategy.
+
+```
+  ...
+  lines = p | beam.io.ReadStringsFromPubSub(topic=known_args.input_topic)
```
Comment (aaltay) [inline src/documentation/sdks/python-streaming.md]: `ReadStringsFromPubSub` is deprecated. (But it is fine to keep it in the doc for now.)

@udim Should we update the example?

File: src/documentation/sdks/python-streaming.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+  ...
+
+  output = counts | 'format' >> beam.Map(format_result)
+
+  # Write to Pub/Sub
+  output | beam.io.WriteStringsToPubSub(known_args.output_topic)
+```
+
+## Running a streaming pipeline
```
Comment (aaltay) [inline src/documentation/sdks/python-streaming.md]: Should we explain how to create input/output pubsub topics and how to publish messages to those? (These are not related to Beam.)

File: src/documentation/sdks/python.md [hunk 6]
```diff
@@ -6,22 +6,26 @@ permalink: /documentation/sdks/python/
...
 
-Get started with the [Beam Programming Guide]({{ site.baseurl }}/documentation/programming-guide) to learn the basic concepts that apply to all SDKs in Beam. Then, follow the [Beam Python SDK Quickstart]({{ site.baseurl }}/get-started/quickstart-py) to set up your Python development environment, get the Beam SDK for Python, and run an example pipeline.
+Get started with the [Beam Python SDK quickstart]({{ site.baseurl }}/get-started/quickstart-py) to set up your Python development environment, get the Beam SDK for Python, and run an example pipeline. Then, read through the [Beam programming guide]({{ site.baseurl }}/documentation/programming-guide) to learn the basic concepts that apply to all SDKs in Beam.
 
-See the [Python API Reference]({{ site.baseurl }}/documentation/sdks/pydoc/) for more information on individual APIs.
+See the [Python API reference]({{ site.baseurl }}/documentation/sdks/pydoc/) for more information on individual APIs.
 
-## Python Type Safety
+## Python streaming pipelines
+
```
Comment (aaltay) [inline src/documentation/sdks/python.md]: TBD?

File: src/get-started/wordcount-example.md [hunk 7]
```diff
@@ -890,24 +890,104 @@ PCollection<KV<String, Long>> wordCounts = windowedWords.apply(new WordCount.Cou
...
+  --streaming
+```
+
+To view the full code in Python, see
+**[streaming_wordcount.py](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/examples/streaming_wordcount.py).**
+
+
+### Reading an unbounded data set
+
+This example uses an unbounded data set as input. THe code reads Pub/Sub
```
Comment (aaltay) [inline src/get-started/wordcount-example.md]: THe -> The

File: src/documentation/sdks/python-streaming.md (hunk 4 above)
Comment (udim) [reply src/documentation/sdks/python-streaming.md]: Yes, using ReadFromPubSub.

**Comment by mariapython** [comment]
In https://github.com/apache/beam-site/blob/asf-site/src/get-started/quickstart-py.md
the line `pip install --upgrade virtualenv` 
should read `pip install --upgrade --user virtualenv` 
to avoid messages of the kind:
   `OSError: [Errno 13] Permission denied: '/usr/local/bin/virtualenv'`

**Comment by udim** [comment]
@mariapython installing packages using --user puts them (by default) under `~/.local/`.
To install `virtualenv` systemwide, we could tell users to `sudo pip install ...`.

**Comment by mariapython** [comment]
@aaltay: Where do the version requirements come from? For example, why do we need pip >= 7.0.0?

**Comment by mariapython** [comment]
On the same file, https://github.com/apache/beam-site/blob/asf-site/src/get-started/quickstart-py.md:
The section on installing pip should actually say something similar to the other install section: "If you do not have... install it by running: pip install --upgrade --user pip"

**Comment by mariapython** [comment]
@udim: Is that a problem? What are the consequences of installing at `~/.local/.`?

**Comment by mariapython** [comment]
On the same file, https://github.com/apache/beam-site/blob/asf-site/src/get-started/quickstart-py.md:
On the section "Execute a pipeline"
- To run wordcount.py, run ... sounds a bit repetitive, I would say "For example, run wordcount.py with the following commend:"
- The command itself should be modified to match the convention above, that is:
`python -m apache_beam.examples.wordcount --input path/to/given/input --output path/to/write/output`
- Add an explanation after the command: “After running the command, a file output-0000-of-0001 will be written to path/to/write/.”
(As a side note, I don't know why we have that notation instead of the more natural <file_name>-x-of-y with x in [1, y])

**Comment by mariapython** [comment]
File https://github.com/apache/beam-site/blob/asf-site/src/get-started/quickstart-py.md, section Download and install --> Extra Requirements:
- Why should "requirements" be capitalized? ("install" is not)
- `pip install apache-beam[feature1, feature2]` doesn't work. It should read `pip install apache-beam[feature1,feature2]` (no space after comma).

**Comment by mariapython** [comment]
File beam-site/src/get-started/wordcount-example.md:
- Section "MinimalWordCount example"
 - Since there is an "Adapt for: Python/Java" at the beginning of the website, the lines To run this example in Java: and "To run this example in Python:" are unnecessary, as the display should toggle them accordingly.
  - In the case of "To run this example in Java:", there should be tabs for every runner, just like in Python. The one currently shown probably is just for the DirectRunner.

**Comment by mariapython** [comment]
(*** This section is incomplete, I will remove this note when I complete its review.)
File src/get-started/mobile-gaming-example.md:
- Section "UserScore: Basic Score Processing in Batch":
  - To be completed
- Section "HourlyTeamScore: Advanced Processing in Batch with Windowing":
  - To be completed
- Section "LeaderBoard: Streaming Processing with Real-Time Game Data":
  - To be completed
- Section "GameStats: Abuse Detection and Usage Analysis":
  - To be completed

**Comment by udim** [comment]
@mariapython Installing packages under `~/.local/` seems to make thing more complicated, but not everyone can (or should) upgrade system packages (because dependencies might break). So perhaps the best way is to use `--user` when installing pip and virtualenv, and install the rest of the packages in a virtualenv to be consistent with our contribution and quickstart guides.

```
$ pip install --user --upgrade pip virtualenv
...
$ python - pip --version
pip 10.0.0 from /home/user/.local/lib/python2.7/site-packages/pip (python 2.7)
$ $HOME/.local/bin/virtualenv --version
...
```

File: src/get-started/wordcount-example.md [hunk 8]
```diff
@@ -890,24 +890,104 @@ PCollection<KV<String, Long>> wordCounts = windowedWords.apply(new WordCount.Cou
...
-need to make sure that we choose an appropriate, unbounded sink. Some output
-sinks support only bounded output, while others support both bounded and
-unbounded outputs. By using a `FilenamePolicy`, we can use `TextIO` to files
-that are partitioned by windows. We use a composite `PTransform` that uses such
-a policy internally to write a single sharded file per window.
+The StreamingWordCount example is a streaming pipeline that reads Pub/Sub
+messages from a Pub/Sub subscription or topic, and performs a frequency count on
+the words in each message. Similar to WindowedWordCount, this example applies
+fixed-time windowing, wherein each window represents a fixed time interval. The
+fixed window size for this example is 15 minutes. The pipeline outputs the
```
Comment (mariapython) [inline src/get-started/wordcount-example.md]: It should read "15 seconds"
(source at `https://github.com/apache/beam/blob/master/sdks/python/apache_beam/examples/streaming_wordcount.py#L77`

File: src/get-started/wordcount-example.md [hunk 9]
```diff
@@ -890,24 +890,104 @@ PCollection<KV<String, Long>> wordCounts = windowedWords.apply(new WordCount.Cou
...
+frequency count of the words seen in each 15 minute window.
```
Comment (mariapython) [inline src/get-started/wordcount-example.md]: Likewise, it should read "15 seconds."

**Comment by melap** [comment]
retest this please

**Comment by melap** [comment]
retest this please

**Comment by aaltay** [comment]
@melap what is the status of this PR? Do you need my help?

File: src/documentation/programming-guide.md [hunk 10]
```diff
@@ -2682,7 +2706,7 @@ example, the following example trigger code fires on the following conditions:
       .withAllowedLateness(Duration.standardDays(2)));
 ```
 ```py
-  # The Beam SDK for Python does not support triggers.
+pcollection | WindowInto(AfterWatermark(late=AfterProcessingTime(10 * 60)))
```
[the remark this answers is not in the document]
Comment (melap) [reply src/documentation/programming-guide.md]: I think the section is more about AfterWatermark, so I'd like to leave something there for Python, but I set the comment about the 2 day allowed lateness only shown if set to Java.

**Comment by melap** [comment]
Made a bunch of updates, PTAL. After it looks good, I'll squash.

**Comment by melap** [comment]
@asfgit merge
